incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -91,9 +104,58 @@ func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Resu
 	// TODO report the "SubnamespaceConflict" in the HNS reconciliation. See issue:
 	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/490
 
+	// Make sure finalizers are set before writing the instance.
+	inst.ObjectMeta.Finalizers = []string{api.MetaGroup}
 	return ctrl.Result{}, r.writeInstance(ctx, log, inst)
 }
 
+func (r *HierarchicalNamespaceReconciler) syncDeleting(log logr.Logger, inst *api.HierarchicalNamespace) (updateFinalizers bool) {
+	r.forest.Lock()
+	defer r.forest.Unlock()
+
+	state := inst.Status.State
+	cnm := inst.Name
+	pnm := inst.Namespace
+	cns := r.forest.Get(cnm)
+	updateFinalizers = false
+
+	switch {
+	// If the owned namespace doesn't exist, remove the finalizer. The missing state
+	// is excluded here in case it's being created by the HC reconciler concurrently.
+	case state == api.Conflict || state == api.Forbidden || !cns.Exists():
+		r.removeFinalizers(log, inst, "the owned namespace doesn't exist")
+		updateFinalizers = true
+	// If the owned namespace exists but doesn't allow cascading deletion, remove the
+	// finalizer and set NNS_MISSING condition in the owned namespace.
+	case !cns.AllowCascadingDelete():
+		r.removeFinalizers(log, inst, "the owned namespace doesn't allow cascading delete")
+		updateFinalizers = true
+		cns.SetCondition(pnm, api.HNSMissing, "missing HNS in the owner namespace")
+		r.hcr.enqueueAffected(log, "updating HNS_MISSING condition", cnm)
+	// If the owned namespace exists and allow cascading deletion, delete it.
+	default:
+		log.Info("setting ns.deleting to true in the forest")
+		cns.SetDeleting()
+		r.hcr.enqueueAffected(log, "cascading deleting this owned namespace from the owner", cnm)
+	}
+
+	return
+}
+
+// removeFinalizers removes the instance's finalizers. If the owner namespace is
+// also being deleted, enqueue the owner to let HC reconciler update its finalizers.
+func (r *HierarchicalNamespaceReconciler) removeFinalizers(log logr.Logger, inst *api.HierarchicalNamespace, reason string) {
+	log.Info("Removing the finalizers", "reason", reason)
+	inst.ObjectMeta.Finalizers = []string{}
+
+	// Only enqueue the owner to update its HC finalizers if the owner is also being deleted.
+	pnm := inst.Namespace
+	pns := r.forest.Get(pnm)
+	if pns.Deleting() {
+		r.hcr.enqueueAffected(log, "updating the finalizers since an HNS instance is deleted", pnm)
+	}
```

Comment:
I'm not sure this belongs in a function called `removeFinalizers` - can you explain? I would have thought we should only enqueue the owner _after_ this object has been deleted, not before...?